src/native/libs/System.Globalization.Native/pal_collation.m

```diff
@@ -45,33 +60,211 @@ static NSStringCompareOptions ConvertFromCompareOptionsToNSStringCompareOptions(
+
+    return modifiedString;
+}
+
+/*
+Function: IndexOf
+Find detailed explanation how this function works in https://github.com/dotnet/runtime/blob/main/docs/design/features/globalization-hybrid-mode.md
+*/
+Range GlobalizationNative_IndexOfNative(const uint16_t* localeName, int32_t lNameLength, const uint16_t* lpTarget, int32_t cwTargetLength,
+                                        const uint16_t* lpSource, int32_t cwSourceLength, int32_t comparisonOptions, int32_t fromBeginning)
+{
+    assert(cwTargetLength >= 0);
+    Range result = {-1, 0};
+    NSStringCompareOptions options = ConvertFromCompareOptionsToNSStringCompareOptions(comparisonOptions);
+    
+    // in case mapping is not found
+    if (options == 0)
+        return result;
+    
+    NSString *searchString = [NSString stringWithCharacters: lpTarget length: cwTargetLength];
+    NSString *searchStrCleaned = RemoveWeightlessCharacters(searchString);
+    NSString *sourceString = [NSString stringWithCharacters: lpSource length: cwSourceLength];
+    NSString *sourceStrCleaned = RemoveWeightlessCharacters(sourceString);
+
+    if (sourceStrCleaned.length == 0 || searchStrCleaned.length == 0)
     {
-        currentLocale = [NSLocale systemLocale];
+       result.location = fromBeginning ? 0 : sourceString.length;
+       return result;
     }
-    else
+
+    NSLocale *currentLocale = GetCurrentLocale(localeName, lNameLength);
+    NSString *searchStrPrecomposed = searchStrCleaned.precomposedStringWithCanonicalMapping;
+    NSString *sourceStrPrecomposed = sourceStrCleaned.precomposedStringWithCanonicalMapping;
+
+    // last index
+    if (!fromBeginning)
+        options |= NSBackwardsSearch;
+
+    // check if source contains search string
+    NSRange rangeOfReceiverToSearch = NSMakeRange(0, sourceStrPrecomposed.length);
+    NSRange containsRange = [sourceStrPrecomposed rangeOfString:searchStrPrecomposed
+                                                  options:options
+                                                  range:rangeOfReceiverToSearch
+                                                  locale:currentLocale];
+
+    if (containsRange.location == NSNotFound)
     {
-        NSString *locName = [NSString stringWithCharacters: localeName length: lNameLength];
-        currentLocale = [[NSLocale alloc] initWithLocaleIdentifier:locName];
+        return result;
+    }
+
+    // localizedStandardRangeOfString is performing a case and diacritic insensitive, locale-aware search and finding first occurance.
+    if ((comparisonOptions & IgnoreCase) && lNameLength == 0 && fromBeginning)
+    {      
+        NSRange localizedStandardRange = [sourceStrCleaned localizedStandardRangeOfString:searchStrCleaned];
+        if (localizedStandardRange.location != NSNotFound)
+        {
+            result.location = localizedStandardRange.location;
+            result.length = localizedStandardRange.length;                    
+            return result;
+        }       
```

Comment:
Migrating comment

`localizedStandardRangeOfString` is performing case and diacritic insensitive, locale-aware search by implementation, that is why I am checking for `IgnoreCase` and if locale is not passed .(see https://developer.apple.com/documentation/foundation/nsstring/1413574-localizedstandardrangeofstring)

----------

Right, I see that it finds the first match with a case insensitive and diacritic insensitive search, but are the other 7 permutations of these options (`!(compareOptions & IgnoreCase) || lNameLength != 0 || !fromBeginning`) covered by the following logic?

I'm wondering if this block is truly necessary. If `((comparisonOptions & IgnoreCase) && lNameLength == 0 && !fromBeginning)` is covered by the following logic, I would imagine that this case is covered as well without this block.

Reply:
Why does `lNameLength == 0` mean locale-aware? If its supposed to be the system locale `[NSLocale systemLocale]`, then shouldn't this check for `localeName == NULL || lNameLength == 0`

Reply:
Wouldn't the `IgnoreNonSpace` be the diacritic insensitive condition as well?

Reply:
`localizedStandardRangeOfString` is not necessary , if it makes the function more complicated I will remove it.

Reply:
I think the block suggested to me that `localizedStandardRangeOfString` could cover one specific case that the rest of the following logic couldn't, which felt odd specifically because that this wouldn't be hit if `fromBeginning` was false. Another consideration is that it seems to ignore the rest of the flags of `options` besides case insensitive and diacritic insensitive, so I wonder if this block would have been correct in those scenarios (if we had the test cases for them). I think it could make sense to keep it if it was much faster than `rangeOfString:options:range:locale:`, but it seems cleaner to rely on the following logic if it can be handled correctly.

Reply:
Now it handles correctly all cases without `localizedStandardRangeOfString`.